### PR TITLE
Modernize Python metadata

### PR DESCRIPTION
**Pull request by mkoeppe**
Using pyproject.toml

### CODE REVIEW

**Comment by mkoeppe** [comment]
I had to comment out the rather unusual `atexit.register(build_json_table)` at the end of `setup.py`. This does not work in builds with build isolation (`pip install .` or `python -m build .`), for example because `click` is not available at build time.

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,55 @@
+[build-system]
+requires = [
+    "setuptools>=61.2",
+]
+build-backend = "setuptools.build_meta"
+
+[project]
+name = "Mathics_Scanner"
+version = "1.3.0"
```
Comment (rocky) [inline pyproject.toml:9]: How can we this read in from `mathics_scanner/version.py?

File: setup.py [hunk 2]
```diff
@@ -145,4 +107,4 @@ def build_json_table() -> int:
     return result.returncode
 
 
-atexit.register(build_json_table)
+# atexit.register(build_json_table)  ## does not work with build isolation
```
Comment (rocky) [inline setup.py:110]: In setup we need to build some tables.

Can this be done in that dynamic section like it is for "version"?

**Comment by rocky** [comment]
@mkoeppe Overall I like this PR and it is needed. To get to 3.12 we'll need do this not just here but also in mathics-core.

Note that the CI failure are not a problem of this PR, but a problem in master. That should now be fixed though.

File: setup.py (hunk 2 above)
Comment (mkoeppe) [reply setup.py:110]: > In setup we need to build some tables.

If users download the sdist (source tarball) from PyPI, would the tables be included there, or would they be built on the user's system?

File: pyproject.toml (hunk 1 above)
Comment (mkoeppe) [reply pyproject.toml]: Thanks. Fixed in 104c114

File: setup.py (hunk 2 above)
Comment (rocky) [reply setup.py:110]: > > In setup we need to build some tables.
> 
> If users download the sdist (source tarball) from PyPI, would the tables be included there, or would they be built on the user's system?

Tables are included in PyPI tarballs, eggs, and wheel. This is done via https://github.com/Mathics3/mathics-scanner/blob/master/setup.py#L102-L104 along with Makefile targets which make sure files are around in development.

Comment (mkoeppe) [reply setup.py]: > I think this was _also_ added to `setup.py` as a way to generate tables as well

Yes, that's the one that I had to comment out.

I've just pushed a change ([52fbe3d](https://github.com/Mathics3/mathics-scanner/pull/58/commits/52fbe3d38e0e1c9d0546269685b9b0db65842cde) that does this in the `egg_info` phase of building an sdist (or wheel)

**Comment by mkoeppe** [comment]
> Note that the CI failure are not a problem of this PR, but a problem in master. That should now be fixed though.

Thanks. I've rebased on top of it

**Comment by rocky** [comment]
Here are diffs to make black happy: 

```diff
index c7efc5b..b2f845a 100755
--- a/mathics_scanner/generate/build_tables.py
+++ b/mathics_scanner/generate/build_tables.py
@@ -15,7 +15,7 @@ try:
     from mathics_scanner.version import __version__
 except ImportError:
     # When using build isolation
-    __version__ = 'unknown'
+    __version__ = "unknown"
 
 
 def get_srcdir():
diff --git a/setup.py b/setup.py
index cf4dfd5..d0c7634 100644
--- a/setup.py
+++ b/setup.py
@@ -86,11 +86,15 @@ class table_building_egg_info(egg_info):
 
     def finalize_options(self):
         """Run program to create JSON tables"""
-        build_tables_program = osp.join(get_srcdir(), "mathics_scanner", "generate", "build_tables.py")
+        build_tables_program = osp.join(
+            get_srcdir(), "mathics_scanner", "generate", "build_tables.py"
+        )
         print(f"Building JSON tables via {build_tables_program}")
         result = subprocess.run([sys.executable, build_tables_program])
         if result.returncode:
-            raise RuntimeError(f"Running {build_tables_program} exited with code {result.returncode}")
+            raise RuntimeError(
+                f"Running {build_tables_program} exited with code {result.returncode}"
+            )
         super().finalize_options()

```

**Comment by rocky** [comment]
Looks Great - Many thanks!

**Pull request merged**